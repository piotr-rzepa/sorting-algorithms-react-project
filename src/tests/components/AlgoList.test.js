import * as React from 'react';
import { shallow } from 'enzyme';
import * as reactRedux from 'react-redux';
import AlgorithmList from '../../components/AlgoList';
import AlgorithmListItem from '../../components/AlgoListItem';
const data = [
	{
		id: 0,
		name: 'Bubble Sort',
		description:
			'Bubble sort, sometimes referred to as sinking sort, is a simple sorting algorithm',
		method: 'Exchanging',
		sortingTime: { best: 0, worst: 0, average: 10 }
	},
	{
		id: 1,
		name: 'Quick Sort',
		description: 'Quicksort is a divide-and-conquer algorithm.',
		method: 'Partitioning',
		sortingTime: { best: 0, worst: 0, average: 100 }
	}
];

describe('<AlgorithmList />', () => {
	const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
	const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
	const useContextMock = jest.spyOn(React, 'useContext');
	beforeEach(() => {
		useDispatchMock.mockClear();
		useSelectorMock.mockClear();
		useContextMock.mockClear();
	});

	it('Should render AlgorithmList correctly', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<AlgorithmList />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should render expected amount of children item components', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<AlgorithmList />);
		expect(wrapper.find(AlgorithmListItem)).toHaveLength(2);
		expect(wrapper.find(AlgorithmListItem).get(0).props.name).toBe('Bubble Sort');
	});

	it('Should render children in correct order (descending)', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<AlgorithmList />);
		expect(wrapper.find(AlgorithmListItem).get(0).props.name).toBe('Bubble Sort');
	});

	it('Should handle change of sorting order to asc', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<AlgorithmList />);
		wrapper
			.find('select')
			.at(0)
			.simulate('change', { target: { value: 'Ascending' } });
		expect(dummyDispatch.mock.calls[0][0]).toEqual({
			type: 'SET_SORT_ASC',
			payload: { sort: 'asc' }
		});
	});

	it('Should handle change of sorting order to desc', () => {
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<AlgorithmList />);
		wrapper
			.find('select')
			.at(0)
			.simulate('change', { target: { value: 'Descending' } });
		expect(dummyDispatch.mock.calls[0][0]).toEqual({
			type: 'SET_SORT_DESC',
			payload: { sort: 'desc' }
		});
	});

	it('should choose correct theme', () => {
		useContextMock.mockReturnValue(true);
		const dummyDispatch = jest.fn();
		useDispatchMock.mockReturnValue(dummyDispatch);
		useSelectorMock.mockReturnValue(data);
		const wrapper = shallow(<AlgorithmList />);
		expect(wrapper.find('div.algorithm-list').get(0).props.style).toHaveProperty(
			'color',
			'white'
		);
	});
});
