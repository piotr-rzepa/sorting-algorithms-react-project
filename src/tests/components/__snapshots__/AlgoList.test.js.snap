// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlgorithmList /> Should render AlgorithmList correctly 1`] = `
<div
  className="algorithm-list"
  style={
    Object {
      "backgroundColor": "white",
      "boxShadow": "2px 2px 2px lightgray",
      "color": "black",
    }
  }
>
  <div
    className="algorithm-list-title"
  >
    <p>
      Algorithms by sorting time
    </p>
    <select
      className="select"
      name="sorting"
      onChange={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "border": "none",
          "color": "lightgray",
        }
      }
    >
      <option>
        Descending
      </option>
      <option>
        Ascending
      </option>
    </select>
  </div>
  <hr />
  <ul>
    <AlgoListItem
      key="0"
      name="Bubble Sort"
      sortingTime={
        Object {
          "average": 10,
          "best": 0,
          "worst": 0,
        }
      }
    />
    <AlgoListItem
      key="1"
      name="Quick Sort"
      sortingTime={
        Object {
          "average": 100,
          "best": 0,
          "worst": 0,
        }
      }
    />
  </ul>
</div>
`;
